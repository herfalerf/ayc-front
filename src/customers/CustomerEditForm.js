import React, { useState, useEffect } from "react";
import { useParams, useHistory } from "react-router-dom";
import { Formik, Form, Field } from "formik";
import { TextField } from "formik-material-ui";
import { Button, ButtonGroup, Grid, Typography } from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";
import ConfirmDialog from "../common/ConfirmDialog";
import * as yup from "yup";
import AycApi from "../api/api";

// Customer edit form
//
// Displays customer edit form
// Submitting the form calls the API to save
//
// Routed as /admin/customer/:id

const useStyles = makeStyles((theme) => ({
  upper: {
    backgroundImage: `url('${process.env.PUBLIC_URL}/images/admin-home/admin-home01.png')`,
    backgroundSize: "cover",
    backgroundPosition: "center bottom",
    minHeight: "25vh",
  },
  form: {
    padding: theme.spacing(4),
    backgroundColor: "rgb(244,245,248)",
  },
  message: {
    color: "#00c7d7",
  },
}));

const CustomerEditForm = ({ editCustomer, deleteCustomer }) => {
  const { id } = useParams();
  console.debug("CustomerEditForm", "id=", id);
  const [customer, setCustomer] = useState({
    name: "",
    email: "",
    phone: "",
    company: "",
  });
  const phoneRegExp = /^(\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]\d{3}[\s.-]\d{4}$/;

  const classes = useStyles();

  const history = useHistory();
  const schema = yup.object().shape({
    name: yup.string().required("Name required"),
    email: yup
      .string()
      .email("Invalid email format")
      .required("Email Required"),
    phone: yup
      .string()
      .matches(phoneRegExp, "Please enter a valid phone number"),
  });

  useEffect(
    function getCustomer() {
      async function getCust() {
        setCustomer(await AycApi.getCustomer(id));
      }
      getCust();
    },
    [id]
  );

  return (
    <div>
      <div className={classes.upper}></div>
      <Grid container justifyContent="center" className={classes.form}>
        <Grid item>
          <Typography variant="h5" className={classes.message}>
            Edit Customer
          </Typography>
          <Formik
            validationSchema={schema}
            initialValues={{
              name: customer.name,
              email: customer.email,
              company: customer.company,
              phone: customer.phone,
            }}
            enableReinitialize={true}
            onSubmit={async (values, { setStatus, setSubmitting }) => {
              setStatus(undefined);
              console.debug("id", id, "values=", values);
              let result = await editCustomer(id, values);
              console.debug(result);

              if (result.success) {
                setSubmitting(false);
                history.push("/admin/customers");
              } else {
                if (
                  result.errors[0] === `Email already exists: ${values.email}`
                ) {
                  setSubmitting(false);
                  history.push("/");
                } else {
                  setStatus({ error: result.errors });
                }
              }
            }}
          >
            {({ isSubmitting, status }) => (
              <Form>
                <Field
                  label="name*"
                  type="name"
                  name="name"
                  component={TextField}
                />
                <br />
                <Field
                  label="email*"
                  type="email"
                  name="email"
                  component={TextField}
                />
                <br />
                <Field
                  label="company"
                  type="company"
                  name="company"
                  component={TextField}
                />
                <br />
                <Field
                  label="phone"
                  type="phone"
                  name="phone"
                  component={TextField}
                />
                <br />
                {status && status.error ? (
                  <div>API Error: {status.error}</div>
                ) : null}
                <br />
                <ButtonGroup>
                  <Button
                    variant="contained"
                    color="primary"
                    disabled={isSubmitting}
                    type="submit"
                  >
                    Submit
                  </Button>
                  <ConfirmDialog
                    name="Delete"
                    title="Are you sure you want to delete?"
                    color="secondary"
                    onConfirm={async () => {
                      let result = await deleteCustomer(id);

                      if (result.success) {
                        history.push("/admin/customers");
                      } else {
                        console.debug("API error", result.errors);
                      }
                    }}
                  ></ConfirmDialog>
                </ButtonGroup>
              </Form>
            )}
          </Formik>
        </Grid>
      </Grid>
    </div>
  );
};

export default CustomerEditForm;
