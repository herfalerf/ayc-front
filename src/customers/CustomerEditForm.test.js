import React from "react";
import { render } from "@testing-library/react";
import CustomerEditForm from "./CustomerEditForm";
import { MemoryRouter } from "react-router-dom";

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <CustomerEditForm />
    </MemoryRouter>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <CustomerEditForm />;
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
