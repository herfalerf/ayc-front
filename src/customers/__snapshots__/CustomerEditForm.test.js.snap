// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="makeStyles-upper-4"
    />
    <div
      class="MuiGrid-root makeStyles-form-5 MuiGrid-container MuiGrid-justify-content-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <h5
          class="MuiTypography-root makeStyles-message-6 MuiTypography-h5"
        >
          Edit Customer
        </h5>
        <form
          action="#"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              name*
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="name"
                type="name"
                value=""
              />
            </div>
          </div>
          <br />
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              email*
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="email"
                type="email"
                value=""
              />
            </div>
          </div>
          <br />
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              company
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="company"
                type="company"
                value=""
              />
            </div>
          </div>
          <br />
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              phone
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="phone"
                type="phone"
                value=""
              />
            </div>
          </div>
          <br />
          <br />
          <div
            class="MuiButtonGroup-root"
            role="group"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined MuiButton-containedPrimary"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-label"
              >
                Submit
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <div>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Delete
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
  ;
</DocumentFragment>
`;
